### ISSUE TITLE
[spec] Phase D -- add GLA (Gated Linear Attention) block as falsifiable arch alternative

### ISSUE THREAD

**Issue by gHashTag** [gHashTag/t27#1004]
## Goal

Add `specs/nn/gla.t27` -- a Gated Linear Attention (GLA) block spec -- as a
CPU-friendly, linear-complexity alternative to the existing full softmax
attention in `specs/nn/attention.t27`. Wire it into the RACE architecture via
an `ATTN_MODE` env-var selector consistent with `arch_config.rs` (Wave 31
PR-B style).

This is a NEW forward-looking spec addition. It does NOT modify existing
`.t27` files. All performance claims are `[Open conjecture]` pending ablation.

---

## Why

The current RACE architecture uses O(N^2) full softmax attention
(`attention.t27`, `model_hybrid_attn.rs`). For the IGLA regime (long-sequence,
CPU inference), linear-complexity attention is a motivated alternative:

- GLA recurrent inference is O(N * d^2) -- constant KV memory, linear in N.
- Yang et al. 2023 (arXiv:2312.06635) demonstrated GLA competitive with
  LLaMA-architecture transformers and faster throughput than Mamba on moderate-
  scale language modeling.
- The RACE trainer already supports env-var arch selection (HIDDEN_DIM,
  NUM_ATTN_LAYERS, GF16_ENABLED). Adding ATTN_MODE="gla" follows the same
  pattern.

The hypothesis that GLA beats full attention at RACE scale is **[Open
conjecture]**. This spec enables the ablation to test it.

---

## Source of truth

- **Spec:** `specs/nn/gla.t27` (created by this issue)
- **Reference:** Yang et al. 2023, "Gated Linear Attention Transformers with
  Hardware-Efficient Training," arXiv:2312.06635,
  https://doi.org/10.48550/arXiv.2312.06635
- **Related specs:** `specs/nn/attention.t27` (full-attn reference);
  `specs/nn/phi_rope.t27` (RoPE, cross-ref for use_rope flag);
  `specs/ml/optimizer/adamw.t27` (optimizer unchanged)
- **Claim-status rules:** `docs/coordination/TASK_PROTOCOL.md`; Law L5 (phi^2
  + phi^-2 = 3 is the ONLY [Verified] phi fact).

---

## Deliverable

1. `specs/nn/gla.t27` -- new module `GatedLinearAttention` with:
   - `GlaConfig`, `GlaHeadState`, `GlaWeights` structs
   - `gla_parallel_forward` (chunk-mode training, O(N*d^2/chunk_size))
   - `gla_state_update` + `gla_state_read` (recurrent O(N*d^2) inference)
   - `apply_gate_act` (Sigmoid / Swish / SoftPlus variants)
   - `gla_config_from_race` (RACE env-var adapter)
   - 14 test blocks, 9 invariant blocks, 5 bench blocks (L4)
2. Trainer: add `ATTN_MODE` env-var parsing to `arch_config.rs` (returns
   `ArchConfig` with new `attn_mode: AttnMode` field). Codegen via t27c.
3. Update `docs/NOW.md` + root `NOW.md` with today's date (now-sync-gate).

---

## Done when

- [ ] `specs/nn/gla.t27` exists with `test`, `invariant`, and `bench` blocks
- [ ] Seal passes: `./scripts/tri seal specs/nn/gla.t27 --save` writes
      `.trinity/seals/gla.json` without error
- [ ] Test suite passes: `./scripts/tri test` includes gla.t27 suite; all
      `test` and `invariant` blocks pass
- [ ] Conformance passes: `./scripts/tri validate-conformance` green (no
      changes to FORMAT-SPEC-001.json)
- [ ] Gen headers valid: `./scripts/tri validate-gen-headers` green
- [ ] `ATTN_MODE` env var is parsed by `arch_config.rs`; `parse_arch_config()`
      returns a struct with `attn_mode` field
- [ ] Ablation matrix documented in PR description: arms GLA-ctrl (gain=1.0),
      GLA-phi (gain=PHI_SQ, [Open conjecture]), Full-attn reference, GLA-rand
- [ ] All phi claims carry `[Open conjecture]` labels; no hype words
- [ ] NOW.md sync-gate green: `./scripts/tri check-now`
- [ ] PR links `Closes #181` and `Closes #<this issue>`

---

## How to verify

```bash
# 1. Seal the new spec
./scripts/tri seal specs/nn/gla.t27 --save

# 2. Run the test suite (includes gla.t27 test/invariant blocks)
./scripts/tri test

# 3. Validate conformance (must not touch FORMAT-SPEC-001.json)
./scripts/tri validate-conformance

# 4. Validate generated headers are consistent
./scripts/tri validate-gen-headers

# 5. Check NOW.md sync gate
./scripts/tri check-now

# 6. Run bench blocks (optional; informational during review)
./scripts/tri bench specs/nn/gla.t27

# 7. Ablation smoke check: verify ATTN_MODE env var is parsed
ATTN_MODE=gla HIDDEN_DIM=256 cargo run --bin trios-train -- --steps 100 --seed 47
# Expect: no panic, finite BPB output, "attn_mode=gla" in log

# 8. Falsification check: run control and phi arms for 3 seeds each
# (full 54k-step ablation is a separate tracking comment on #181)
```

---

## Risks

1. **State stability:** The GLA matrix state (HEAD_DIM x HEAD_DIM) can grow
   unbounded if gate values approach 1.0 indefinitely. The normaliser floor
   (`max(|z.q|, 1.0)`) partially mitigates this, but long-sequence stability
   must be verified empirically. Mitigation: add `is_finite` assertion in the
   `gla_state_update_preserves_finiteness` invariant and monitor during
   training.

2. **Complexity crossover:** GLA is theoretically O(N * d^2) vs O(N^2 * d)
   for full attention. At RACE default HEAD_DIM=64, the crossover is ~N=64
   tokens. At RACE default seq_len the advantage may be marginal. The bench
   blocks (`gla_full_pass_seq64`, `gla_full_pass_seq512`) will quantify this.

3. **phi gain interaction:** Setting qk_gain=PHI_SQ_GAIN in GLA may interact
   differently than in full attention (Phase B2 INV-9) because the GLA output
   is normalised by the z-dot-q denominator. The phi ablation must be run
   separately for GLA vs full-attn.

4. **L6 CEILING:** The spec deliberately avoids touching
   `conformance/FORMAT-SPEC-001.json`. If the codegen pipeline requires a new
   format entry for GLA state dtype, that must go through a separate L6-aware
   PR.

5. **Trainer coupling:** The RACE trainer (`trios-train`) must be updated to
   read `ATTN_MODE`. This is trainer code (not t27 gen/), so it requires a
   separate PR to `gHashTag/trios-trainer-igla`.

---

## Links

- Parent epic: #181 (phi->IGLA)
- Full-attn spec: `specs/nn/attention.t27`
- Phase B2 (QK-Gain): `specs/nn/attention.t27` INV-9
- GLA paper: Yang et al. 2023, arXiv:2312.06635,
  https://doi.org/10.48550/arXiv.2312.06635
- Hybrid stretch goal (D2): Gated DeltaNet Yang 2024, arXiv:2412.06464,
  https://doi.org/10.48550/arXiv.2412.06464; Kimi Linear, arXiv:2510.26692;
  GSA recall fix Zhang 2024, arXiv:2409.07146,
  https://doi.org/10.48550/arXiv.2409.07146
- RACE arch config: `igla-peek/src/arch_config.rs` (Wave 31 PR-B)
- FLASHLINEARATTENTION reference: Yang 2023 Algorithm 1 (same paper)
- Falsification methodology: Melis/Dyer/Blunsom 2017 equal-budget tuning
  (arXiv:1707.05589); Li & Talwalkar 2019 random-search baseline
